### PR TITLE
[inductor] Change AOTInductor to return output tensors

### PR DESCRIPTION
**Pull request by desertfire**
Summary:
Change AOTInductor to directly return output tensors instead of taking pre-allocated output tensors to return the results. This gives several benefits:

* It makes sure AOTInductor has the same behavior when managing the output tensors as the default Inductor, which is widely tested and thus more reliable.
* As we have debugged before, there are cases we still have to codegen extra copy_ ops to fill the pre-allocated output tensors which doesn't make sense for performance.
* With the coming enhanced memory planning, this again will make sure the memory planning logic is the between AOTInductor and Inductor, which will greatly simplify the problem and improve the reliability.

This change also combines D49494954 from Yang and https://github.com/pytorch/pytorch/pull/109560 from Angela.

Differential Revision: D49502318

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/csrc/inductor/aot_runtime/model.h [hunk 1]
```diff
@@ -364,6 +341,8 @@ class AOTInductorModelBase {
 
   std::vector<ParamInfo> inputs_info_;
   std::vector<ParamInfo> outputs_info_;
+  std::vector<AtenTensorHandle> output_handles_;
```
Comment (desertfire) [inline torch/csrc/inductor/aot_runtime/model.h:344]: Didn't remove this. Will do in the next PR update.

File: test/cpp/aot_inductor/test.cpp [hunk 2]
```diff
@@ -55,46 +55,36 @@ TEST(AotInductorTest, BasicTest) {
...
-  AOTI_RUNTIME_ERROR_CODE_CHECK(AOTInductorModelContainerGetMaxOutputShape(
-      container_handle, 0 /*output_idx*/, &max_output_sizes, &max_output_dim));
 
-  c10::IntArrayRef array_size(max_output_sizes, max_output_dim);
-  torch::Tensor output_tensor =
-      at::zeros(array_size, at::dtype(at::kFloat).device(at::kCUDA));
-  std::vector<torch::Tensor> outputs;
-  outputs.push_back(output_tensor);
+  auto input_handles =
+      torch::aot_inductor::unsafe_alloc_new_handles_from_tensors(input_tensors);
```
Comment (chenyang78) [inline test/cpp/aot_inductor/test.cpp:60]: Hmm, do we need to change the var name at line 48 to ```input_tensors```?

File: test/cpp/aot_inductor/test.cpp [hunk 3]
```diff
@@ -55,46 +55,36 @@ TEST(AotInductorTest, BasicTest) {
...
-      output_sizes.data(),
-      output_ndims.data()));
-
-  ASSERT_EQ(output_sizes.size(), 1);
-  ASSERT_EQ(output_ndims[0], 2);
-  ASSERT_EQ(output_sizes[0][0], 32);
-  ASSERT_EQ(output_sizes[0][1], 10);
+      proxy_executor_handle));
+
+  outputs = torch::aot_inductor::alloc_tensors_by_stealing_from_handles(
```
Comment (chenyang78) [inline test/cpp/aot_inductor/test.cpp:85]: Unless I am missing the place, seems we don't declare outputs yet? If that's the case, perhaps change it to ```auto outputs```?

File: torch/_inductor/codegen/wrapper.py [hunk 4]
```diff
@@ -1207,7 +1195,7 @@ def write_wrapper_decl(self):
                     c10::optional<c10::string_view> optional_string;
                     c10::optional<at::Layout> optional_layout;
                     c10::optional<at::Tensor> optional_tensor;
-                    torch::List<c10::optional<at::Scalar>> optional_list;
+                    c10::List<c10::optional<at::Scalar>> optional_list;
```
Comment (chenyang78) [inline torch/_inductor/codegen/wrapper.py:1198]: dumb question - why are we switching to ```c10::List```?

**Review by chenyang78 — APPROVED** [review]
Left some minor comments. LGTM. Thanks!

File: test/cpp/aot_inductor/test.cpp (hunk 2 above)
Comment (desertfire) [reply test/cpp/aot_inductor/test.cpp:60]: Good catch. I should have run this cpp test locally.

File: test/cpp/aot_inductor/test.cpp (hunk 3 above)
Comment (desertfire) [reply test/cpp/aot_inductor/test.cpp:85]: Running this cpp test locally.

File: torch/_inductor/codegen/wrapper.py (hunk 4 above)
Comment (desertfire) [reply torch/_inductor/codegen/wrapper.py:1198]: It was coming my other include file change. Just for uniformity.

**Pull request closed without merging**